### PR TITLE
Stepper component

### PR DESCRIPTION
**Pull request by cameronviner**
#### What does this PR do? (please provide any background)

Adds a stepper component to to the UI Toolkit

#### What tests does this PR have?

Unit tests added to cover the new component

#### How can this be tested?

checkout the branch and then run:

`grunt docs`

scroll to the bottom of the page and test the steppers functionality. 

#### Screenshots / Screencast

Stepper image

#### What gif best describes how you feel about this work?

---
#### Developer Definition of Done/Quality Checklist (for PR author to complete BEFORE code review):
- [x] I have checked the [contributing document](../blob/master/CONTRIBUTING.md) and I'm happy for this to be reviewed.

#### Software Engineer or Developer review:
- [x] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [x] I’ve checked for coding anti-patterns.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked all the tests are passing.

#### Software Engineer or project guru review:
- [x] I’ve witnessed the work behaving as expected (this could be on the authors machine or screencast).
- [x] I’ve checked for coding anti-patterns.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked all the tests are passing.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,18 +38,21 @@
...
   "homepage": "https://github.com/holidayextras/ui-toolkit",
   "dependencies": {
-    "browserify": "11.1.0",
+    "browserify": "^11.1.0",
     "classnames": "^2.1.0",
     "font-awesome": "^4.3.0",
     "grunt": "^0.4.5",
     "grunt-shell": "^1.1.1",
     "gruntfile-gtx": "^0.3.0",
+    "jest-cli": "^0.5.10",
```
Comment (jodiedoubleday) [inline package.json]: Why has Jest been introduced again?

Comment (cameronviner) [inline package.json]: Not something I remember adding in, I have removed the dependency.

**Comment by hxtomprice** [comment]
Dev

**Comment by hxtomprice** [comment]
I'm getting the same issue as before where the stepper loses the event handler. This time it only seems to happen when you hit the limiters

File: docs/src/Components.jsx [hunk 2]
```diff
@@ -225,6 +225,20 @@ var Components = React.createClass({
             <small>We currently use a small sample of <a href="http://www.alessioatzeni.com/meteocons/" title="Meteocons">Meteocons</a></small>
           </article>
 
+          <article id="stepper">
+            <h3>Stepper</h3>
+            <p>Display an input with a number value that is surrounded by decrement and increment buttons</p>
+            <CustomComponent codeText={fs.readFileSync(__dirname + '/../examples/Stepper.jsx', 'utf8')} />
+            <h4>Attributes</h4>
+            <ul>
+              <li><code>value</code> Number - The value of the stepper</li>
```
Comment (hxtomprice) [inline docs/src/Components.jsx]: Should this read "The initial value of the stepper"?

File: package.json [hunk 3]
```diff
@@ -47,9 +47,11 @@
     "lodash.pick": "~3.1.0",
     "lodash.values": "~3.0.0",
     "moment": "~2.10.6",
-    "react": "^0.13.2",
+    "react": "^0.13.3",
     "react-data-attributes-mixin": "git://github.com/holidayextras/react-data-attributes-mixin",
-    "reactify": "~1.1.0",
+    "react-tools": "^0.13.3",
+    "reactify": "^1.1.1",
+    "redirectify": "^1.4.0",
```
Comment (hxtomprice) [inline package.json]: `redirectify` and `react-tools` have been added but aren't used?

**Comment by hxtomprice** [comment]
@cameronviner suggested removing `node_modules` from the `docs/` directory and installing again. Stepper is all working as expected now

**Comment by hxpaul** [comment]
taking a review

File: docs/src/Components.jsx (hunk 2 above)
Comment (hxpaul) [inline docs/src/Components.jsx]: I think this is ok, the onChange call will re-render the stepper passing in a new actual value...

**Comment by lukehansell** [comment]
Trading have a new requirement for their wizard, a stepper which displays a value with text (eg "2am"). How difficult would it be to extend this to match this new use case? I'm not suggesting we do it here, merely asking your thoughts.

File: dist/ui-toolkit.standalone.js [hunk 4]
```diff
@@ -28436,10 +28435,104 @@ module.exports = require('./code/index');
...
+      value: 0,
+      incrementDisplayString: '+',
+      decrementDisplayString: '-',
+      onChange: function() {}
+    };
+  },
+
+  getInitialState: function() {
+    return {
+      value: this.props.value
```
Comment (hxpaul) [inline dist/ui-toolkit.standalone.js]: this one is an antipattern, I have been reliably informed! I think we can get away without the stepper maintaining its own state at all. Only use this.props.value and do not increment it but call the onChange with this.props.value + 1 or this.props.value - 1. Let me check this with someone with more react experience though.

Comment (lukehansell) [inline dist/ui-toolkit.standalone.js]: :+1:

File: dist/ui-toolkit.standalone.js [hunk 5]
```diff
@@ -28436,10 +28435,104 @@ module.exports = require('./code/index');
...
+    };
+  },
+
+  decrement: function() {
+    if (!this.canDecrement()) return;
+    var value = this.state.value - 1;
+    this.setState({ value: value });
+    this.props.onChange(this.state.value);
```
Comment (hxpaul) [inline dist/ui-toolkit.standalone.js]: I think we can simplify this, get rid of the state and just `this.props.onChange( this.props.value - 1 )` but I'm going to check this now.
Also I think we might want to put in a step size option, though as that's not a requirement now we should leave that until it is.

File: dist/ui-toolkit.standalone.js (hunk 4 above)
Comment (hxpaul) [inline dist/ui-toolkit.standalone.js]: consider that checked! Get rid of state completely here, it simplifies your code. Then one more general point, if you run coverage you'll see if your new test covers all the possibilities - right now it does not.
```
npm run coveralls;
open coverage/lcov-report/index.html 
```
some of the other tests do not cover enough either, that's not your problem and I will fix that, but you should aim for 100% coverage with your stepper test. When you get rid of state you will find you have less code to cover so your test might even get simpler.
`coveralls` is the name of the npm script here though it's likely to change to `npm run coverage` soon if we standardise...

**Comment by hxpaul** [comment]
to cover displaying an alternative value to the actual number you're incrementing (2am, 5 cakes, etc) you might pass in an optional transform function... easy enough I think, but shouldn't bog down this release imo

**Comment by cameronviner** [comment]
@hxpaul @hxtomprice I've updated this now with your changes and better test coverage (an empty function is stopping this getting 100%). Could you take another look when you get a chance?

File: package.json [hunk 6]
```diff
@@ -48,9 +48,9 @@
     "lodash.pick": "~3.1.0",
     "lodash.values": "~3.0.0",
     "moment": "~2.10.6",
-    "react": "^0.13.2",
+    "react": "^0.13.3",
     "react-data-attributes-mixin": "git://github.com/holidayextras/react-data-attributes-mixin",
-    "reactify": "~1.1.0",
+    "redirectify": "^1.4.0",
```
Comment (hxpaul) [inline package.json]: don't think this is needed is it?

**Comment by hxpaul** [comment]
it is a thing of beauty now :+1:

**Comment by hxtomprice** [comment]
:+1:

**Comment by hxpaul** [comment]
Seen it working too, I'm using it in the hotel wizard now...

**Pull request merged**